mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -151,4 +153,77 @@ class StringIndexerModel private[ml] (
     val copied = new StringIndexerModel(uid, labels)
     copyValues(copied, extra)
   }
+
+  /**
+   * Return a model to perform the inverse transformation.
+   * Note: By default we keep the original columns during this transformation, so the inverse
+   * should only be used on new columns such as predicted labels.
+   */
+  def invert(inputCol: String, outputCol: String): StringIndexerInverseTransformer = {
+    val labelsCol: String = $(this.outputCol)
+    new StringIndexerInverseTransformer(labelsCol)
+      .setInputCol(inputCol)
+      .setOutputCol(outputCol)
+  }
+}
+
+/**
+ * :: Experimental ::
+ * Transform a provided column back to the original input types using the metadata on the
+ * labelsCol. Note: By default we keep the original columns during this transformation,
+ * so the inverse should only be used on new columns such as predicted labels.
+ */
+@Experimental
+class StringIndexerInverseTransformer private[ml] (
+  override val uid: String,
+  val labelsCol: String) extends Transformer
```

Comment:
Well, it should be...but it's true that many Models do not update the output columns appropriately right now.

Maybe we should go back closer to what you had before, where StringIndexerInverse takes a StringArrayParam called "labels" which lets the user provide the lookup table?  If it is not specified, then the input column must have the metadata.  If it is specified, then the input column metadata are ignored.  StringIndexerModel.invert() can set this parameter.  Later on, it may become unnecessary, but that's OK.